File: ros/gs130_ros/test/test_convert.py
"""Tests for the conversions from gs130 values to ROS messages."""

import math

import numpy as np
import pytest
from builtin_interfaces.msg import Time

import gs130
from gs130_ros import convert


# ---------------------------------------------------------------------------
# A stand-in for gs130.Image: the same shape and the same two properties, so
# the packing tests below exercise the real slicing arithmetic.
# ---------------------------------------------------------------------------


class _Frame(np.ndarray):
    def __new__(cls, array):
        return np.asarray(array, dtype=np.uint8).view(cls)

    @property
    def width(self):
        return self.shape[1]

    @property
    def height(self):
        return self.shape[0] * 2 // 3


class _Intrinsics:
    """The fields convert reads off a gs130.CameraIntrinsics."""

    def __init__(self, fx, fy, cx, cy, dist_model=gs130.DistModel.PINHOLE, dist=None):
        self.fx = fx
        self.fy = fy
        self.cx = cx
        self.cy = cy
        self.K = np.array([[fx, 0.0, cx], [0.0, fy, cy], [0.0, 0.0, 1.0]])
        self.dist_coeffs = np.zeros(8) if dist is None else np.asarray(dist, dtype=float)
        self.dist_model = dist_model


class _Packet:
    def __init__(self, accel, gyro, timestamp_ns):
        self.accel = np.asarray(accel, dtype=float)
        self.gyro = np.asarray(gyro, dtype=float)
        self.temp = 25.0
        self.is_fsync = True
        self.timestamp_ns = timestamp_ns


def _rotation_from_quaternion(x, y, z, w):
    return np.array(
        [
            [1 - 2 * (y * y + z * z), 2 * (x * y - z * w), 2 * (x * z + y * w)],
            [2 * (x * y + z * w), 1 - 2 * (x * x + z * z), 2 * (y * z - x * w)],
            [2 * (x * z - y * w), 2 * (y * z + x * w), 1 - 2 * (x * x + y * y)],
        ]
    )


# ---------------------------------------------------------------------------
# Timestamps
# ---------------------------------------------------------------------------


def test_to_time_splits_seconds_and_nanoseconds():
    moment = convert.to_time(1_500_000_002)
    assert (moment.sec, moment.nanosec) == (1, 500_000_002)
    assert (convert.to_time(0).sec, convert.to_time(0).nanosec) == (0, 0)


def test_to_time_keeps_nanoseconds_in_range_for_negative_values():
    # ROS allows a negative sec as long as nanosec stays in [0, 1e9).
    moment = convert.to_time(-1)
    assert (moment.sec, moment.nanosec) == (-1, 999_999_999)


def test_to_time_refuses_a_value_past_the_ros_range():
    with pytest.raises(ValueError):
        convert.to_time((2**31) * 1_000_000_000)


# ---------------------------------------------------------------------------
# Rotation
# ---------------------------------------------------------------------------


@pytest.mark.parametrize("axis", [0, 1, 2])
@pytest.mark.parametrize("angle", [0.0, 30.0, 90.0, 179.0, 180.0, 270.0])
def test_quaternion_round_trips_through_its_matrix(axis, angle):
    radians = math.radians(angle)
    rotation = np.eye(3)
    rotation[(axis + 1) % 3, (axis + 1) % 3] = math.cos(radians)
    rotation[(axis + 2) % 3, (axis + 2) % 3] = math.cos(radians)
    rotation[(axis + 1) % 3, (axis + 2) % 3] = -math.sin(radians)
    rotation[(axis + 2) % 3, (axis + 1) % 3] = math.sin(radians)

    quaternion = convert.quaternion(rotation)
    assert math.isclose(sum(value * value for value in quaternion), 1.0, abs_tol=1e-12)
    # q and -q are the same rotation, so compare the matrices they build.
    assert np.allclose(_rotation_from_quaternion(*quaternion), rotation, atol=1e-9)


def test_quaternion_refuses_a_non_finite_rotation():
    with pytest.raises(ValueError):
        convert.quaternion(np.full((3, 3), np.nan))


# ---------------------------------------------------------------------------
# Projection
# ---------------------------------------------------------------------------


def test_projection_has_no_baseline_before_rectification():
    projection = convert.projection(_Intrinsics(600.0, 600.0, 544.0, 640.0))
    assert projection[3] == 0.0
    assert projection[7] == 0.0
    assert projection[11] == 0.0


def test_stereo_projection_puts_the_baseline_in_p3():
    intrinsics = _Intrinsics(600.0, 600.0, 544.0, 640.0)
    # The right eye sits a baseline to the left of the left eye's frame.
    projection = convert.stereo_projection(intrinsics, (-0.08, 0.0, 0.0))
    assert projection[3] == pytest.approx(600.0 * -0.08)
    assert projection[7] == 0.0
    assert projection[11] == 0.0
    # What a stereo matcher reads back out of it.
    assert abs(projection[3] / projection[0]) == pytest.approx(0.08)


def test_stereo_projection_accounts_for_an_off_axis_offset():
    intrinsics = _Intrinsics(600.0, 600.0, 544.0, 640.0)
    projection = convert.stereo_projection(intrinsics, (-0.08, 0.0, 0.002))
    assert projection[3] == pytest.approx(600.0 * -0.08 + 544.0 * 0.002)


def test_stereo_projection_of_the_left_eye_has_no_baseline():
    projection = convert.stereo_projection(
        _Intrinsics(600.0, 600.0, 544.0, 640.0), (0.0, 0.0, 0.0)
    )
    assert projection[3] == 0.0


# ---------------------------------------------------------------------------
# CameraInfo
# ---------------------------------------------------------------------------


def test_camera_info_declares_the_eye_size_and_an_identity_r():
    message = convert.camera_info(
        _Intrinsics(600.0, 601.0, 544.0, 640.0), 1088, 1280, "camera_link", Time()
    )
    assert (message.width, message.height) == (1088, 1280)
    assert list(message.k[:3]) == [600.0, 0.0, 544.0]
    assert list(message.r) == [1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0]
    assert len(message.p) == 12


def test_camera_info_carries_all_eight_pinhole_coefficients():
    intrinsics = _Intrinsics(600.0, 600.0, 544.0, 640.0, dist=np.arange(8, dtype=float))
    message = convert.camera_info(intrinsics, 1088, 1280, "camera_link", Time())
    assert message.distortion_model == "rational_polynomial"
    assert list(message.d) == list(range(8))


def test_camera_info_carries_four_coefficients_for_a_fisheye():
    intrinsics = _Intrinsics(
        600.0, 600.0, 544.0, 640.0, gs130.DistModel.FISHEYE, dist=np.arange(8, dtype=float)
    )
    message = convert.camera_info(intrinsics, 1088, 1280, "camera_link", Time())
    assert message.distortion_model == "equidistant"
    assert list(message.d) == [0.0, 1.0, 2.0, 3.0]


def test_a_rectified_frame_needs_no_distortion_model():
    intrinsics = _Intrinsics(
        600.0, 600.0, 544.0, 640.0, gs130.DistModel.FISHEYE, dist=np.arange(8, dtype=float)
    )
    message = convert.camera_info(
        intrinsics, 1088, 1280, "camera_link", Time(), rectified=True
    )
    assert message.distortion_model == "plumb_bob"
    assert list(message.d) == [0.0] * 5


# ---------------------------------------------------------------------------
# Packed frames
# ---------------------------------------------------------------------------


@pytest.mark.parametrize(
    "layout",
    [
        gs130.StereoLayout.NONE,
        gs130.StereoLayout.TOP_BOTTOM,
        gs130.StereoLayout.BOTTOM_TOP,
        gs130.StereoLayout.LEFT_RIGHT,
        gs130.StereoLayout.RIGHT_LEFT,
    ],
)
def test_eye_size_is_the_eye_not_the_frame(layout):
    eye_width, eye_height = 128, 96
    frame_width, frame_height = _frame_size(layout, eye_width, eye_height)
    assert convert.eye_size(layout, frame_width, frame_height) == (eye_width, eye_height)


def _frame_size(layout, eye_width, eye_height):
    """The frame the C layer builds around an eye of the given size."""
    if layout in (gs130.StereoLayout.TOP_BOTTOM, gs130.StereoLayout.BOTTOM_TOP):
        return eye_width, eye_height * 2
    if layout in (gs130.StereoLayout.LEFT_RIGHT, gs130.StereoLayout.RIGHT_LEFT):
        return eye_width * 2, eye_height
    return eye_width, eye_height


def _nv12(width, height, fill):
    return np.full(width * height * 3 // 2, fill, dtype=np.uint8)


def _bands(nv12, width, height):
    """Split one NV12 image into its luma rows and its chroma rows."""
    rows = nv12.reshape(-1, width)
    return rows[:height], rows[height:]


def _packed_frame(layout, width, height):
    """Build the buffer gs130.cpp writes for ``layout`` (gs130.cpp:262-301).

    Written from the C offsets, not from ``slice_eyes``: top/bottom puts the
    first eye's luma at 0, the second's at ``w*h``, then their chroma planes at
    ``2*w*h`` and ``2*w*h + w*h/2``; left/right interleaves the two eyes inside
    every row, at a stride of ``2*w``.
    """
    left = _nv12(width, height, 0x11)
    right = _nv12(width, height, 0x22)

    left_luma, left_chroma = _bands(left, width, height)
    right_luma, right_chroma = _bands(right, width, height)

    left_first = layout in (gs130.StereoLayout.TOP_BOTTOM, gs130.StereoLayout.LEFT_RIGHT)
    first = (left_luma, left_chroma) if left_first else (right_luma, right_chroma)
    second = (right_luma, right_chroma) if left_first else (left_luma, left_chroma)

    if layout in (gs130.StereoLayout.TOP_BOTTOM, gs130.StereoLayout.BOTTOM_TOP):
        # Luma of both eyes, then chroma of both: the two planes are separate
        # buffers handed to the hardware, not one image appended to another.
        packed = np.concatenate([first[0], second[0], first[1], second[1]])
    else:
        packed = np.concatenate(
            [
                np.concatenate([first[0], second[0]], axis=1),
                np.concatenate([first[1], second[1]], axis=1),
            ]
        )
    packed = packed.reshape(-1)

    frame_width, frame_height = _frame_size(layout, width, height)
    return _Frame(packed.reshape(frame_height * 3 // 2, frame_width))


@pytest.mark.parametrize(
    "layout",
    [
        gs130.StereoLayout.TOP_BOTTOM,
        gs130.StereoLayout.BOTTOM_TOP,
        gs130.StereoLayout.LEFT_RIGHT,
        gs130.StereoLayout.RIGHT_LEFT,
    ],
)
def test_slice_eyes_recovers_both_eyes(layout):
    width, height = 8, 6
    packed = _packed_frame(layout, width, height)
    left, right = convert.slice_eyes(packed, layout)
    eye_size = width * height * 3 // 2
    assert left.size == eye_size and right.size == eye_size
    assert np.all(left == 0x11)
    assert np.all(right == 0x22)


def test_slice_eyes_returns_arrays_that_outlive_the_frame():
    width, height = 8, 6
    packed = _packed_frame(gs130.StereoLayout.TOP_BOTTOM, width, height)
    left, right = convert.slice_eyes(packed, gs130.StereoLayout.TOP_BOTTOM)
    packed.fill(0)
    assert np.all(left == 0x11)
    assert np.all(right == 0x22)


# ---------------------------------------------------------------------------
# Images, IMU, transforms
# ---------------------------------------------------------------------------


def test_nv12_message_describes_the_buffer_it_carries():
    width, height = 8, 6
    frame = _Frame(_nv12(width, height, 0x33).reshape(height * 3 // 2, width))
    message = convert.nv12_message(frame, width, height, "camera_link", Time(sec=7))
    assert message.encoding == "nv12"
    assert (message.width, message.height, message.step) == (width, height, width)
    assert len(message.data) == width * height * 3 // 2
    assert message.header.frame_id == "camera_link"
    assert message.header.stamp.sec == 7


def test_nv12_message_payload_is_an_array_not_a_bytes():
    """The payload type is a performance contract, not a style choice.

    rclpy fills a ``uint8[]`` field one element at a time when it is given a
    ``bytes``: 3.3 seconds for a 1088x2560 NV12 frame on an RDK X5, against
    0.8 ms for an ``array.array``.  Nothing else about the two differs, so only
    this assertion stops the slow one coming back.
    """
    import array

    width, height = 8, 6
    source = _nv12(width, height, 0x33)
    frame = _Frame(source.reshape(height * 3 // 2, width))
    message = convert.nv12_message(frame, width, height, "camera_link", Time())
    assert isinstance(message.data, array.array)
    assert message.data.typecode == "B"
    assert bytes(message.data) == source.tobytes()


def test_nv12_message_copies_so_the_frame_can_be_dropped():
    width, height = 8, 6
    frame = _Frame(_nv12(width, height, 0x33).reshape(height * 3 // 2, width))
    message = convert.nv12_message(frame, width, height, "camera_link", Time())
    frame.fill(0)
    assert bytes(message.data) == bytes([0x33]) * (width * height * 3 // 2)


def test_imu_message_marks_the_missing_orientation():
    message = convert.imu_message(
        _Packet([1.0, 2.0, 3.0], [4.0, 5.0, 6.0], 9), "imu_link", Time()
    )
    acceleration = message.linear_acceleration
    assert (acceleration.x, acceleration.y, acceleration.z) == (1.0, 2.0, 3.0)
    rate = message.angular_velocity
    assert (rate.x, rate.y, rate.z) == (4.0, 5.0, 6.0)
    assert message.orientation_covariance[0] == -1.0


def test_transform_message_passes_the_device_pose_through():
    rotation = np.eye(3)
    message = convert.transform_message(
        rotation, (1.0, 2.0, 3.0), "camera_link", "imu_link", Time()
    )
    assert message.header.frame_id == "camera_link"
    assert message.child_frame_id == "imu_link"
    translation = message.transform.translation
    assert (translation.x, translation.y, translation.z) == (1.0, 2.0, 3.0)
    rotation = message.transform.rotation
    assert (rotation.x, rotation.y, rotation.z, rotation.w) == (0.0, 0.0, 0.0, 1.0)

File: ros/gs130_ros/test/test_convert.cpp
// Copyright (c) 2026 D-Robotics.
// SPDX-License-Identifier: MIT
//
// Tests for the conversions from gs130 values to ROS messages.
//
// These link only gs130_ros_convert and the generated messages, never
// libgs130: the conversions use the C header's types and call none of it, so
// they can be checked on a machine with no hardware and no library.

#include <gtest/gtest.h>

#include <array>
#include <cmath>
#include <cstdint>
#include <cstring>
#include <vector>

#include "gs130_ros/convert.hpp"

namespace gs130_ros
{
namespace
{

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

gs130_camera_intrinsics_t intrinsics(
  double fx, double fy, double cx, double cy,
  gs130_dist_model_t model = GS130_DIST_PINHOLE, double distortion = 0.0)
{
  gs130_camera_intrinsics_t values{};
  values.fx = fx;
  values.fy = fy;
  values.cx = cx;
  values.cy = cy;
  values.dist_model = model;
  values.K[0] = fx;
  values.K[2] = cx;
  values.K[4] = fy;
  values.K[5] = cy;
  values.K[8] = 1.0;
  for (int i = 0; i < 8; ++i) {
    values.dist_coeffs[i] = distortion == 0.0 ? 0.0 : distortion + i;
  }
  return values;
}

std::array<double, 9> rotation_about(int axis, double degrees)
{
  const double radians = degrees * M_PI / 180.0;
  const double c = std::cos(radians);
  const double s = std::sin(radians);
  std::array<double, 9> rotation{1, 0, 0, 0, 1, 0, 0, 0, 1};
  const int a = (axis + 1) % 3;
  const int b = (axis + 2) % 3;
  rotation[a * 3 + a] = c;
  rotation[b * 3 + b] = c;
  rotation[a * 3 + b] = -s;
  rotation[b * 3 + a] = s;
  return rotation;
}

std::array<double, 9> matrix_from_quaternion(const double q[4])
{
  const double x = q[0], y = q[1], z = q[2], w = q[3];
  return {
    1 - 2 * (y * y + z * z), 2 * (x * y - z * w), 2 * (x * z + y * w),
    2 * (x * y + z * w), 1 - 2 * (x * x + z * z), 2 * (y * z - x * w),
    2 * (x * z - y * w), 2 * (y * z + x * w), 1 - 2 * (x * x + y * y)};
}

std::vector<uint8_t> nv12(uint32_t width, uint32_t height, uint8_t fill)
{
  return std::vector<uint8_t>(static_cast<size_t>(width) * height * 3 / 2, fill);
}

/// Build the buffer gs130.cpp writes for `layout` (gs130.cpp:262-301).
///
/// Written from the C offsets, not from slice_eyes, so the two are checked
/// against each other: top/bottom puts both eyes' luma first, at 0 and w*h,
/// then their chroma at 2*w*h and 2*w*h + w*h/2; left/right interleaves the two
/// eyes inside every row, at a stride of 2*w.
std::vector<uint8_t> packed_frame(
  gs130_stereo_layout_t layout, uint32_t width, uint32_t height)
{
  const std::vector<uint8_t> left = nv12(width, height, 0x11);
  const std::vector<uint8_t> right = nv12(width, height, 0x22);
  const size_t luma = static_cast<size_t>(width) * height;
  const size_t chroma = luma / 2;
  const bool left_first = layout == GS130_STEREO_LAYOUT_TOP_BOTTOM ||
    layout == GS130_STEREO_LAYOUT_LEFT_RIGHT;

  std::vector<uint8_t> packed;
  if (layout == GS130_STEREO_LAYOUT_TOP_BOTTOM || layout == GS130_STEREO_LAYOUT_BOTTOM_TOP) {
    const std::vector<uint8_t> & first = left_first ? left : right;
    const std::vector<uint8_t> & second = left_first ? right : left;
    packed.insert(packed.end(), first.begin(), first.begin() + luma);
    packed.insert(packed.end(), second.begin(), second.begin() + luma);
    packed.insert(packed.end(), first.begin() + luma, first.end());
    packed.insert(packed.end(), second.begin() + luma, second.end());
    return packed;
  }

  const std::vector<uint8_t> & first = left_first ? left : right;
  const std::vector<uint8_t> & second = left_first ? right : left;
  for (size_t row = 0; row < luma / width; ++row) {
    packed.insert(packed.end(), first.begin() + row * width, first.begin() + (row + 1) * width);
    packed.insert(packed.end(), second.begin() + row * width, second.begin() + (row + 1) * width);
  }
  for (size_t row = 0; row < chroma / width; ++row) {
    packed.insert(
      packed.end(), first.begin() + luma + row * width, first.begin() + luma + (row + 1) * width);
    packed.insert(
      packed.end(), second.begin() + luma + row * width,
      second.begin() + luma + (row + 1) * width);
  }
  return packed;
}

// ---------------------------------------------------------------------------
// Timestamps
// ---------------------------------------------------------------------------

TEST(ToTime, SplitsSecondsAndNanoseconds)
{
  const auto stamp = to_time(1500000002);
  EXPECT_EQ(stamp.sec, 1);
  EXPECT_EQ(stamp.nanosec, 500000002u);
  EXPECT_EQ(to_time(0).sec, 0);
}

TEST(ToTime, KeepsNanosecondsInRangeForNegativeValues)
{
  // ROS allows a negative sec as long as nanosec stays in [0, 1e9).
  const auto stamp = to_time(-1);
  EXPECT_EQ(stamp.sec, -1);
  EXPECT_EQ(stamp.nanosec, 999999999u);
}

TEST(ToTime, RefusesAValuePastTheRosRange)
{
  EXPECT_THROW(to_time(static_cast<int64_t>(2147483648LL) * 1000000000LL), std::out_of_range);
}

// ---------------------------------------------------------------------------
// Rotation
// ---------------------------------------------------------------------------

TEST(Quaternion, RoundTripsThroughItsMatrix)
{
  for (int axis = 0; axis < 3; ++axis) {
    for (double degrees : {0.0, 30.0, 90.0, 179.0, 180.0, 270.0}) {
      const std::array<double, 9> rotation = rotation_about(axis, degrees);
      double quaternion[4] = {0, 0, 0, 0};
      ASSERT_TRUE(quaternion_from_rotation(rotation.data(), quaternion))
        << "axis " << axis << " degrees " << degrees;
      const double norm = std::sqrt(
        quaternion[0] * quaternion[0] + quaternion[1] * quaternion[1] +
        quaternion[2] * quaternion[2] + quaternion[3] * quaternion[3]);
      EXPECT_NEAR(norm, 1.0, 1e-12);
      // q and -q are the same rotation, so compare the matrices they build.
      const std::array<double, 9> rebuilt = matrix_from_quaternion(quaternion);
      for (int i = 0; i < 9; ++i) {
        EXPECT_NEAR(rebuilt[i], rotation[i], 1e-9) << "axis " << axis << " index " << i;
      }
    }
  }
}

TEST(Quaternion, RefusesANonFiniteRotation)
{
  std::array<double, 9> rotation;
  rotation.fill(std::nan(""));
  double quaternion[4] = {0, 0, 0, 0};
  EXPECT_FALSE(quaternion_from_rotation(rotation.data(), quaternion));
}

// ---------------------------------------------------------------------------
// Projection
// ---------------------------------------------------------------------------

TEST(Projection, HasNoBaselineBeforeRectification)
{
  const auto projection = k_projection(intrinsics(600.0, 600.0, 544.0, 640.0));
  EXPECT_DOUBLE_EQ(projection[3], 0.0);
  EXPECT_DOUBLE_EQ(projection[7], 0.0);
  EXPECT_DOUBLE_EQ(projection[11], 0.0);
}

TEST(Projection, PutsTheBaselineInP3)
{
  const auto values = intrinsics(600.0, 600.0, 544.0, 640.0);
  // relative_T(LEFT, RIGHT): the right eye sits a baseline to the left of the
  // left eye's frame, so tx is negative and P[3] is too.
  const double translation[3] = {-0.08, 0.0, 0.0};
  const auto projection = stereo_projection(values, translation);
  EXPECT_NEAR(projection[3], 600.0 * -0.08, 1e-12);
  EXPECT_DOUBLE_EQ(projection[7], 0.0);
  EXPECT_DOUBLE_EQ(projection[11], 0.0);
  // What a stereo matcher reads back out of it.
  EXPECT_NEAR(std::fabs(projection[3] / projection[0]), 0.08, 1e-12);
}

TEST(Projection, AccountsForAnOffAxisOffset)
{
  const auto values = intrinsics(600.0, 600.0, 544.0, 640.0);
  const double translation[3] = {-0.08, 0.0, 0.002};
  const auto projection = stereo_projection(values, translation);
  EXPECT_NEAR(projection[3], 600.0 * -0.08 + 544.0 * 0.002, 1e-12);
}

TEST(Projection, LeftEyeHasNoBaseline)
{
  const auto values = intrinsics(600.0, 600.0, 544.0, 640.0);
  const double zero[3] = {0.0, 0.0, 0.0};
  EXPECT_DOUBLE_EQ(stereo_projection(values, zero)[3], 0.0);
}

TEST(Projection, MatchesTheMeasuredDevice)
{
  // Measured on an RDK X5 with a GS130WI under rect: relative_T(LEFT, RIGHT)
  // is (-0.07031615052675907, 0, 0), fx = fy = 615.5139302922461, cx = 544.
  const auto values = intrinsics(615.5139302922461, 615.5139302922461, 544.0, 640.0);
  const double translation[3] = {-0.07031615052675907, 0.0, 0.0};
  const auto projection = stereo_projection(values, translation);
  EXPECT_NEAR(projection[3], -43.28057017374667, 1e-9);
  EXPECT_NEAR(std::fabs(projection[3] / projection[0]), 0.07031615052675907, 1e-12);
}

// ---------------------------------------------------------------------------
// CameraInfo
// ---------------------------------------------------------------------------

TEST(CameraInfo, DeclaresTheEyeSizeAndAnIdentityR)
{
  const auto message = camera_info(
    intrinsics(600.0, 601.0, 544.0, 640.0), 1088, 1280, "camera_link", to_time(0), nullptr, false);
  EXPECT_EQ(message.width, 1088u);
  EXPECT_EQ(message.height, 1280u);
  EXPECT_DOUBLE_EQ(message.k[0], 600.0);
  EXPECT_DOUBLE_EQ(message.k[2], 544.0);
  const std::array<double, 9> identity{1, 0, 0, 0, 1, 0, 0, 0, 1};
  for (int i = 0; i < 9; ++i) {
    EXPECT_DOUBLE_EQ(message.r[i], identity[i]);
  }
  EXPECT_EQ(message.p.size(), 12u);
}

TEST(CameraInfo, CarriesAllEightPinholeCoefficients)
{
  const auto message = camera_info(
    intrinsics(600.0, 600.0, 544.0, 640.0, GS130_DIST_PINHOLE, 0.5), 1088, 1280,
    "camera_link", to_time(0), nullptr, false);
  EXPECT_EQ(message.distortion_model, "rational_polynomial");
  EXPECT_EQ(message.d.size(), 8u);
  EXPECT_DOUBLE_EQ(message.d[0], 0.5);
  EXPECT_DOUBLE_EQ(message.d[7], 7.5);
}

TEST(CameraInfo, CarriesFourCoefficientsForAFisheye)
{
  const auto message = camera_info(
    intrinsics(600.0, 600.0, 544.0, 640.0, GS130_DIST_FISHEYE, 0.5), 1088, 1280,
    "camera_link", to_time(0), nullptr, false);
  EXPECT_EQ(message.distortion_model, "equidistant");
  EXPECT_EQ(message.d.size(), 4u);
  EXPECT_DOUBLE_EQ(message.d[3], 3.5);
}

TEST(CameraInfo, ARectifiedFrameNeedsNoDistortionModel)
{
  const auto message = camera_info(
    intrinsics(600.0, 600.0, 544.0, 640.0, GS130_DIST_FISHEYE, 0.5), 1088, 1280,
    "camera_link", to_time(0), nullptr, true);
  EXPECT_EQ(message.distortion_model, "plumb_bob");
  EXPECT_EQ(message.d.size(), 5u);
  EXPECT_DOUBLE_EQ(message.d[0], 0.0);
}

// ---------------------------------------------------------------------------
// Packed frames
// ---------------------------------------------------------------------------

TEST(EyeSize, IsTheEyeNotTheFrame)
{
  uint32_t width = 0;
  uint32_t height = 0;
  eye_size(GS130_STEREO_LAYOUT_NONE, 128, 96, &width, &height);
  EXPECT_EQ(width, 128u);
  EXPECT_EQ(height, 96u);
  eye_size(GS130_STEREO_LAYOUT_TOP_BOTTOM, 128, 192, &width, &height);
  EXPECT_EQ(width, 128u);
  EXPECT_EQ(height, 96u);
  eye_size(GS130_STEREO_LAYOUT_BOTTOM_TOP, 128, 192, &width, &height);
  EXPECT_EQ(width, 128u);
  EXPECT_EQ(height, 96u);
  eye_size(GS130_STEREO_LAYOUT_LEFT_RIGHT, 256, 96, &width, &height);
  EXPECT_EQ(width, 128u);
  EXPECT_EQ(height, 96u);
  eye_size(GS130_STEREO_LAYOUT_RIGHT_LEFT, 256, 96, &width, &height);
  EXPECT_EQ(width, 128u);
  EXPECT_EQ(height, 96u);
}

TEST(SliceEyes, RecoversBothEyesInEveryLayout)
{
  const uint32_t eye_width = 8;
  const uint32_t eye_height = 6;
  const size_t eye_bytes = static_cast<size_t>(eye_width) * eye_height * 3 / 2;

  for (auto layout : {
      GS130_STEREO_LAYOUT_TOP_BOTTOM, GS130_STEREO_LAYOUT_BOTTOM_TOP,
      GS130_STEREO_LAYOUT_LEFT_RIGHT, GS130_STEREO_LAYOUT_RIGHT_LEFT})
  {
    const bool stacked = layout == GS130_STEREO_LAYOUT_TOP_BOTTOM ||
      layout == GS130_STEREO_LAYOUT_BOTTOM_TOP;
    const uint32_t width = stacked ? eye_width : eye_width * 2;
    const uint32_t height = stacked ? eye_height * 2 : eye_height;
    const std::vector<uint8_t> packed = packed_frame(layout, eye_width, eye_height);
    ASSERT_EQ(packed.size(), static_cast<size_t>(width) * height * 3 / 2);

    const auto eyes = slice_eyes(packed.data(), width, height, layout);
    ASSERT_EQ(eyes[0].size(), eye_bytes);
    ASSERT_EQ(eyes[1].size(), eye_bytes);
    EXPECT_TRUE(std::all_of(eyes[0].begin(), eyes[0].end(), [](uint8_t v) {return v == 0x11;}))
      << "left eye wrong for layout " << static_cast<int>(layout);
    EXPECT_TRUE(std::all_of(eyes[1].begin(), eyes[1].end(), [](uint8_t v) {return v == 0x22;}))
      << "right eye wrong for layout " << static_cast<int>(layout);
  }
}

// ---------------------------------------------------------------------------
// Images, IMU, transforms
// ---------------------------------------------------------------------------

TEST(ImageMessage, DescribesTheBufferItCarries)
{
  const uint32_t width = 8;
  const uint32_t height = 6;
  const std::vector<uint8_t> source = nv12(width, height, 0x33);
  const auto message = nv12_message(source.data(), width, height, "camera_link", to_time(0));
  EXPECT_EQ(message.encoding, "nv12");
  EXPECT_EQ(message.width, width);
  EXPECT_EQ(message.height, height);
  EXPECT_EQ(message.step, width);
  EXPECT_EQ(message.data.size(), source.size());
  EXPECT_EQ(message.header.frame_id, "camera_link");
}

TEST(ImageMessage, CopiesSoTheFrameCanBeReleased)
{
  const uint32_t width = 8;
  const uint32_t height = 6;
  std::vector<uint8_t> source = nv12(width, height, 0x33);
  const auto message = nv12_message(source.data(), width, height, "camera_link", to_time(0));
  std::fill(source.begin(), source.end(), 0);
  EXPECT_EQ(message.data[0], 0x33);
  EXPECT_EQ(message.data.back(), 0x33);
}

TEST(ImuMessage, MarksTheMissingOrientation)
{
  gs130_imu_packet_t packet{};
  packet.accel[0] = 1.0f;
  packet.accel[1] = 2.0f;
  packet.accel[2] = 3.0f;
  packet.gyro[0] = 4.0f;
  packet.gyro[1] = 5.0f;
  packet.gyro[2] = 6.0f;
  const auto message = imu_message(packet, "imu_link", to_time(0));
  EXPECT_FLOAT_EQ(message.linear_acceleration.x, 1.0f);
  EXPECT_FLOAT_EQ(message.linear_acceleration.z, 3.0f);
  EXPECT_FLOAT_EQ(message.angular_velocity.y, 5.0f);
  EXPECT_DOUBLE_EQ(message.orientation_covariance[0], -1.0);
}

TEST(TransformMessage, PassesTheDevicePoseThrough)
{
  const std::array<double, 9> rotation{1, 0, 0, 0, 1, 0, 0, 0, 1};
  const double translation[3] = {1.0, 2.0, 3.0};
  const auto message = transform_message(
    rotation.data(), translation, "camera_link", "imu_link", to_time(0));
  EXPECT_EQ(message.header.frame_id, "camera_link");
  EXPECT_EQ(message.child_frame_id, "imu_link");
  EXPECT_DOUBLE_EQ(message.transform.translation.x, 1.0);
  EXPECT_DOUBLE_EQ(message.transform.translation.z, 3.0);
  EXPECT_DOUBLE_EQ(message.transform.rotation.w, 1.0);
}

}  // namespace
}  // namespace gs130_ros
